mountpoint-s3/src/main.rs

```diff
@@ -479,3 +484,38 @@ fn parse_perm_bits(perm_bit_str: &str) -> Result<u16, anyhow::Error> {
         Ok(perm)
     }
 }
+
+pub fn retrieve_instance_type() -> Option<String> {
+    let imds_crt_client = ImdsCrtClient::new()
+        .map_err(|e| {
+            tracing::error!("failed to create instance metadata service client, moutpoint-s3 is unable to auto detect ec2 instance which is currently running on: {:?}", e);
+        }).ok()?;
+
+    let output = futures::executor::block_on(async move {
+        let query = imds_crt_client
+            .make_instance_type_query(move |result| {
+                if let Err(e) = result {
+                    Err(ImdsQueryRequestError::CrtError(e))
+                } else {
+                    Ok(result)
+                }
+            })
+            .map_err(|e| {
+                tracing::error!("failed to retrieve ec2 instance type from asynchronous channel {:?}", e);
+                e
+            })?;
+        query.await
+    });
+
+    let output = output
+        .map_err(|e| {
+            tracing::error!(
+                "failed to retrieve ec2 instance type by calling AWS Common Runtime(CRT) library {:?}",
+                e
+            );
+            e
+        })
+        .ok()?;
+
+    output.ok()
```

Comment:
If we get the `Result` types fixed up, this can all just be something like
```rust
    let imds_crt_client = ImdsCrtClient::new().context("failed to create IMDS client")?;

    let query = imds_crt_client
        .make_instance_type_query()
        .context("failed to send IMDS query")?;

    Ok(futures::executor::block_on(query).context("IMDS query failed")?)
```

Reply:
fixed